parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java

```diff
@@ -463,14 +486,29 @@ ConvertedType convertToConvertedType(LogicalTypeAnnotation logicalTypeAnnotation
     }
   }
 
-  private void addRowGroup(ParquetMetadata parquetMetadata, List<RowGroup> rowGroups, BlockMetaData block) {
+  private void addRowGroup(ParquetMetadata parquetMetadata, List<RowGroup> rowGroups, BlockMetaData block,
```

Comment:
Should we keep original addRowGroup() intact and just add a new one with InternalFileEncryptor? There is not much duplicate code if doing so and we can refactor the existing code with helper functions.

Reply:
encryption is isolated there, with _if (null != fileEncryptor)_  and _if (encryptMetaData)_ - similar to the _if (columnIndexRef != null)_  and _if (offsetIndexRef != null)_ in the same function.

Reply:
Not all the changes are isolated. Generally, adding 'if/else' will add diverge the code and add the complexity. One other thing is regression thinking. If fileEncryptor is null, which would be most of the case, then it just executes the existing method without change. It would be less error prone.

Reply:
Many functions can be run either with or without encryption. Duplicating them will result in hundreds or thousands of duplicate code lines. This will make code maintenance (changes/fixes)  a headache. Instead, we isolate encryption with _if_ switches, without duplicating the existing code.

The same goes for other recent new features (column indexes and bloom filters) - they are isolated with an _if_ switch, instead of code duplication. See _if (columnIndexRef != null) and if (offsetIndexRef != null)_ in this addRowGroup function.

Reply:
It is not a blocking comment and I am fine with it. But generally speaking, adding too much nested if/else diverges the code path and causes the complexity for reading. One way to avoid duplicating is to wrap them up in helper functions. I understand column indexes and bloom filters already did that but if we keep adding features like this, the code will become less and less readable.